### ISSUE TITLE
initDataLoader only initializes datasetSize/batchSize indices, causing wrong batches for batchSize > 1

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#74]
## Summary

`initDataLoader` and `dataLoaderInit` disagree on the semantics of the `indices` array. The allocation sizes it for one entry per sample, but the initialization loop only fills `datasetSize / batchSize` entries. For `batchSize > 1`, calls to `getBatch(dl, i)` for `i >= numberOfIndices / batchSize` return the wrong samples — silently, because `reserveMemory` is calloc-backed and the uninitialized indices are zero.

## Location

- `src/userApi/data_loader/DataLoaderApi.c:28-29` — allocates `datasetSize` entries
- `src/data_loader/DataLoader.c:36-44` — initializes and shuffles only `datasetSize / batchSize` entries
- `src/userApi/data_loader/DataLoaderApi.c:54-57` — `getSampleByIndex` treats `indices` as one-entry-per-sample

## Trace

For `datasetSize = 4`, `batchSize = 2`:

**`dataLoaderInit`** (`DataLoaderApi.c:28-29`):
```c
size_t numberOfIndices = getDatasetSize();              // 4
size_t *indices = *reserveMemory(numberOfIndices * sizeof(size_t));  // calloc, all zeros
```

**`initDataLoader`** (`DataLoader.c:36-44`):
```c
size_t sizeDataset = getDatasetSize();                   // 4
size_t numberOfIndices = sizeDataset / dataLoader->batchSize;  // 2  ← different variable, same name
for (size_t i = 0; i < numberOfIndices; ++i)
    indices[i] = i;                                       // indices[0]=0, indices[1]=1
// indices[2], indices[3] remain 0 from calloc

if (shuffle) rngShuffleIndices(indices, numberOfIndices); // shuffles only first 2
```

**`getBatch(dl, 1)`** (`DataLoaderApi.c:59-73`) with `batchSize=2`:
```c
size_t sampleIndex = 1 * 2 = 2;
batch->samples[0] = getSampleByIndex(dl, 2);  // reads indices[2]=0 → getSample(0)
batch->samples[1] = getSampleByIndex(dl, 3);  // reads indices[3]=0 → getSample(0)
```

Returns sample 0 twice instead of samples 2 and 3. `getBatch(dl, 0)` still works correctly.

## Impact

- All multi-batch training/evaluation paths silently produce wrong results when `batchSize > 1` and `datasetSize > batchSize`.
- Shuffle is also broken: only the first `datasetSize / batchSize` indices participate.
- Every test in `test/unit/userAPI/UnitTestTrainingLoopApi.c` uses `batchSize = 1` with the comment "to avoid initDataLoader index initialization bug" as a workaround.
- This blocks writing multi-batch tests for evaluation metrics (#70) and likely any other feature that should be exercised across multiple batches.
- `MnistExperiment.c` uses `batchSize = 32` for training — training may be using degraded batches where only batch 0 is correct.

## Root Cause

The `indices` array is intended to be sample-level (one entry per sample index, supporting per-sample shuffling), but `initDataLoader` treats it as batch-level. The `numberOfIndices` computation in `DataLoader.c:37` is wrong — it should be `sizeDataset`, not `sizeDataset / batchSize`.

## Fix

In `src/data_loader/DataLoader.c`:

```c
size_t sizeDataset = getDatasetSize();

for (size_t i = 0; i < sizeDataset; ++i)
    indices[i] = i;

if (shuffle) {
    rngSetSeed(shuffleSeed);
    rngShuffleIndices(indices, sizeDataset);
}
```

## Test Plan

- [ ] Add a test that constructs a dataloader with `datasetSize = 4`, `batchSize = 2`, calls `getBatch(dl, 0)` and `getBatch(dl, 1)`, and verifies all four samples appear across the two batches.
- [ ] Add a test with `shuffle = true` verifying that all `datasetSize` positions participate in shuffling.
- [ ] Once fixed, remove the `batchSize = 1` workaround comments from `UnitTestTrainingLoopApi.c` and add multi-batch variants of the evaluation/training tests.

## Discovered While

Working on #70 (classification metrics). The evaluation tests had to use `batchSize = 1` exclusively, leaving the multi-batch accumulation path in `evaluateBatchInternal` untested.

**Comment by LeoBuron** [comment]
## Root Cause Clarification

The bug stems from conflating **batch-level** and **sample-level** indexing in `initDataLoader`.

The `indices` array is designed as a **sample-level** mapping — `getSampleByIndex` reads `indices[sampleIndex]` where `sampleIndex` ranges from `0` to `datasetSize - 1`. But `initDataLoader` initializes it as **batch-level**:

```c
size_t numberOfIndices = sizeDataset / dataLoader->batchSize;  // thinks in batches
for (size_t i = 0; i < numberOfIndices; ++i)
    indices[i] = i;  // writes batch numbers, not sample IDs
```

This is conceptually a microbatch/batch confusion. The framework uses microbatch=1 by design (each forward+backward processes one sample for memory efficiency on MCUs). The dataloader's `batchSize` controls how many microbatches are grouped before `optimizer.step()`. But `initDataLoader` treats `batchSize` as if it changes the granularity of `indices`, when in fact `indices` is always sample-granular.

The shuffle has the same problem: `rngShuffleIndices(indices, numberOfIndices)` only shuffles `datasetSize/batchSize` entries instead of all `datasetSize` entries.

### Impact on MnistExperiment

`MnistExperiment.c` uses `batchSize=32` for training with `datasetSize=60000`:
- `numberOfIndices = 60000/32 = 1875` → only `indices[0..1874]` initialized
- `getBatch(dl, 0)`: `sampleIndex=0`, reads `indices[0..31]` → values `0..31` ✓
- `getBatch(dl, 58)`: `sampleIndex=1856`, reads `indices[1856..1887]` → `indices[1875..1887]` are zero from calloc → returns sample 0 repeatedly
- `getBatch(dl, 59+)`: all zero → 32× sample 0 per batch

Only ~1856 out of 60000 training samples are actually used per epoch. All published MnistExperiment results were trained on degraded data.

(Evaluation uses `batchSize=1`, so eval metrics are unaffected.)
